spec/parsers/iso_base_media_file_format/decoder_spec.rb

```diff
@@ -0,0 +1,242 @@
+        expect(result.length).to eq(1)
+        expect(io.pos).to eq(0x14)
+
+        foo_atom = result[0]
+        expect(foo_atom.type).to eq('foo ')
+        expect(foo_atom.position).to eq(0)
+        expect(foo_atom.size).to eq(0x14)
+        expect(foo_atom.fields).to eq({})
+        expect(foo_atom.children).to eq([])
+      end
+    end
+
+    context 'when parsing a container atom' do
+      let(:io) do
+        # moov
+        # |-> foo
+        # |-> bar
+        input = [0x18].pack('N') + 'moov' + [0x8].pack('N') + 'foo ' + [0x8].pack('N') + 'bar '
+        StringIO.new(input)
+      end
+
+      it 'parses type, position, size and children' do
+        result = subject.build_atom_tree(0xFF, io)
+        expect(result.length).to eq(1)
+        expect(io.pos).to eq(0x18)
+
+        moov_atom = result[0]
+        expect(moov_atom.type).to eq('moov')
+        expect(moov_atom.position).to eq(0)
+        expect(moov_atom.size).to eq(0x18)
+        expect(moov_atom.fields).to eq({})
+        expect(moov_atom.children.length).to eq(2)
+      end
+    end
+
+    context 'when parsing an empty atom' do
+      let(:io) do
+        # nmhd
+        # |-> foo
+        input = [0x18].pack('N') + 'nmhd' + [0x1].pack('c') + 'fla' + [0x2].pack('N') + [0x8].pack('N') + 'foo '
+        StringIO.new(input)
+      end
+
+      it 'parses type, position, size, version and flags, and skips any other fields or children' do
+        result = subject.build_atom_tree(0xFF, io)
+        expect(result.length).to eq(1)
+        expect(io.pos).to eq(0x18)
+
+        nmhd_atom = result[0]
+        expect(nmhd_atom.type).to eq('nmhd')
+        expect(nmhd_atom.position).to eq(0)
+        expect(nmhd_atom.size).to eq(0x18)
+        expect(nmhd_atom.fields).to include({
+          version: 1,
+          flags: 'fla'
+        })
+        expect(nmhd_atom.children).to eq([])
+      end
+    end
+
+    context 'when parsing a uuid atom' do
+      let(:usertype) { '90f7c66ec2db476b977461e796f0dd4b' }
+      let(:io) do
+        input = [0x20].pack('N') + 'uuid' + [usertype].pack('H*') + [0x8].pack('N') + 'foo '
+        StringIO.new(input)
+      end
+
+      it 'parses type, position, size and usertype, and skips any other fields or children' do
+        # uuid
+        # |-> foo
+        result = subject.build_atom_tree(0xFF, io)
+        expect(result.length).to eq(1)
+        expect(io.pos).to eq(0x20)
+
+        nmhd_atom = result[0]
+        expect(nmhd_atom.type).to eq('uuid')
+        expect(nmhd_atom.position).to eq(0)
+        expect(nmhd_atom.size).to eq(0x20)
+        expect(nmhd_atom.fields).to include({
+          usertype: usertype,
+        })
+        expect(nmhd_atom.children).to eq([])
+      end
+    end
+  end
+end
+
+describe FormatParser::ISOBaseMediaFileFormat::Decoder::Atom do
+  context 'when initialized' do
+    context 'without fields and/or children' do
+      subject { FormatParser::ISOBaseMediaFileFormat::Decoder::Atom.new('foo', 0, 0) }
```

Comment:
You can make things a bit more bearable with:

```suggestion
      subject { described_class.new('foo', 0, 0) }
```

Reply:
Love it! Thanks!